### PR TITLE
Add tabs for multiple commands

### LINKED ISSUE #10372: Regexp match error when mutiple slashes
## Steps to reproduce

I found a problem when developing cmd_*_tabs method.

```
str = "HKCU\\\\Software"
tab_word = "HKCU\\\\Software\\\\7-Zip"
tab_word  =~ /^#{str}/i  => nil
```

Relevant problem code in `lib/rex/ui/text/dispatcher_shell.rb` line 390 

```
    # Verify that our search string is a valid regex
    begin
      Regexp.compile(str,Regexp::IGNORECASE)
    rescue RegexpError
      str = Regexp.escape(str)
    end
......
    items.find_all { |e|
      e =~ /^#{str}/i
    # Prepend the rest of the command (or it all gets replaced!)
    }.map { |e|
      tab_words.dup.push(e).join(' ')
    }

```

When the `str = "HKCU\\\\Software"`, `tab_word = "HKCU\\\\Software\\\\7-Zip" `:

```
tab_word =~ Regexp.compile(str,Regexp::IGNORECASE) => nil
tab_word =~ Regexp.escape(str) => 0
```
` Regexp.compile(str,Regexp::IGNORECASE) ` would not throw exception, `str = Regexp.escape(str)` would not be executed either.

So I think there is a issue in the regexp match code in `lib/rex/ui/text/dispatcher_shell.rb`, especially on windows. 

### Metasploit version

Framework: 5.0.0-dev-00a748a92c
Console  : 5.0.0-dev-00a748a92c

### I installed Metasploit with:
- [ ] Kali package via apt
- [ ] Omnibus installer (nightly)
- [ ] Commercial/Community installer (from http://www.rapid7.com/products/metasploit/download.jsp)
- [x] Source install (please specify ruby version)

`ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux]`

### OS

`Linux kali 4.6.0-kali1-amd64 #1 SMP Debian 4.6.4-1kali1 (2016-07-21) x86_64 GNU/Linux`

### PR DESCRIPTION
**Pull request by Green-m**
Add tabs for many commands in msfconsole and meterpreter.
Fixes #10372 to get better experience when use tab to auto complete on windows.

## Verification

List the steps needed to make sure this thing works

In `msfconsole`
```
msf5 exploit(multi/handler) > cd ./
cd ./app            cd ./data           cd ./docker         cd ./external       cd ./modules        cd ./script         cd ./spec           cd ./tools          
cd ./config         cd ./db             cd ./documentation  cd ./lib            cd ./plugins        cd ./scripts        cd ./test           

msf5 exploit(multi/handler) > cd ./lib/
cd ./lib/anemone     cd ./lib/metasploit  cd ./lib/net         cd ./lib/rabal       cd ./lib/rex         cd ./lib/sqlmap      cd ./lib/telephony   
cd ./lib/metasm      cd ./lib/msf         cd ./lib/postgres    cd ./lib/rbmysql     cd ./lib/snmp        cd ./lib/tasks       

```

In `meterpreter`
```
msf5 exploit(multi/handler) > sessions -i 4
[*] Starting interaction with 4...

meterpreter > cat c:\\
cat c:\\$Recycle.Bin                      cat c:\\PerfLogs                          cat c:\\Ruby23-x64                        cat c:\\p.csv
cat c:\\477317de0dbfd0e5e29cbb5d9b831a4e  cat c:\\Program Files                     cat c:\\System Volume Information         cat c:\\pagefile.sys
cat c:\\Config.Msi                        cat c:\\Program Files (x86)               cat c:\\Users                             cat c:\\sysmon
cat c:\\DRIVERS                           cat c:\\ProgramData                       cat c:\\Windows                           cat c:\\temp
cat c:\\Documents and Settings            cat c:\\Python27                          cat c:\\cygwin64                          
cat c:\\Intel                             cat c:\\Python35                          cat c:\\ea39c3858c367d4594967f            
cat c:\\MSOCache                          cat c:\\Recovery                          cat c:\\hiberfil.sys                      

```

Hope you guys like this!

### CODE REVIEW

**Comment by wvu-r7** [comment]
Awesome initiative! Thanks so much. :)

**Comment by wvu-r7** [comment]
Hey, @Green-m. Check this out: https://help.github.com/articles/closing-issues-using-keywords/. I think your syntax is pretty close but a little off. I've edited the PR description to close the issue automatically when this is landed. :)

**Comment by Green-m** [comment]
@wvu-r7 Thanks! 
BTW, this doc https://github.com/rapid7/metasploit-framework/wiki/Meterpreter-Transport-Control is out of date : )

**Comment by wvu-r7** [comment]
@Green-m: Do you have a patch I can apply and credit you for?

**Comment by Green-m** [comment]
@wvu-r7 
Yeah, I'm glad to do it!

**Comment by wvu-r7** [comment]
@Green-m: Btw, I didn't see a good way to contact you outside your weibo. I thank you for your contributions to Metasploit. I'm sure the rest of the team at R7 would agree. :)

**Comment by Green-m** [comment]
@wvu-r7 I am in the msf slack, or email acceptable ?

**Comment by wvu-r7** [comment]
@Green-m: I'm not in that Slack. Send it to msfdev@metasploit.com. Thanks!

File: lib/msf/ui/console/command_dispatcher/core.rb [hunk 1]
```diff
@@ -176,6 +176,18 @@ def cmd_color(*args)
...
+  #
+  # Tab completion for the color command
+  #
+  # @param str [String] the string currently being typed before tab was hit
+  # @param words [Array<String>] the previously completed words on the command line.  words is always
+  # at least 1 when tab completion has reached this stage since the command itself has been completed
+
+  def cmd_color_tabs(str, words)
+    return [] if words.length > 1
+    ['auto','true','false']
```
Comment (wvu-r7) [inline lib/msf/ui/console/command_dispatcher/core.rb]: Consider using `%w[]` for your string arrays. A bit more readable for tab completion code.

**Comment by Green-m** [comment]
@wvu-r7 
I have updated this [doc](https://github.com/rapid7/metasploit-framework/wiki/Meterpreter-Transport-Control) for `transport`, please check it.

**Comment by OJ** [comment]
Wow. Great docs 👍

>

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/fs.rb [hunk 2]
```diff
@@ -271,6 +278,13 @@ def cmd_cd(*args)
     return true
   end
 
+  #
+  # Tab completion for the cd command
+  #
+  def cmd_cd_tabs(str, words)
```
Comment (wvu-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/fs.rb]: This is my favorite one. I wish we hadn't waited so long to fix it!

**Comment by wvu-r7** [comment]
It appears that #10433 created a merge conflict here. Sorry!

**Comment by Green-m** [comment]
@wvu-r7 
That's all right. Since the disk of my working PC has crashed, I could only do things on a remote cloud server, which makes my work difficult.

After fixed issue #10499, I would resolve the conflict.

**Comment by Green-m** [comment]
@wvu-r7 
Have resolved conflict, sorry for the late!

**Comment by wvu-r7** [comment]
This is my primary sprint task right now, so I'll get this done before the end of the week!

**Comment by wvu-r7** [comment]
Please see https://github.com/rapid7/metasploit-framework/commit/e955e8dc1530ea44119945de11cbe9d3f8ee3014 and check if I missed or messed up anything. Thanks!

**Comment by wvu-r7** [comment]
# Release Notes

This adds new tab completion functionality to many commands in the console and Meterpreter.

**Pull request merged**